llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp

```diff
@@ -3388,75 +3373,91 @@ SDValue NVPTXTargetLowering::LowerFormalArguments(
   //     individually present in Ins.
   // So a different index should be used for indexing into Ins.
   // See similar issue in LowerCall.
-  unsigned InsIdx = 0;
+  const auto *In = Ins.begin();
+  auto ConsumeArgIns = [&](const Argument &Arg) {
+    const auto *ArgInsBegin = In;
+    const auto *ArgInsEnd = In;
+    while (ArgInsEnd != Ins.end() && ArgInsEnd->OrigArgIndex == Arg.getArgNo())
+      ++ArgInsEnd;
+    In = ArgInsEnd;
+    return llvm::ArrayRef(ArgInsBegin, ArgInsEnd);
+  };
```

Comment:
Would this work?
```
ArrayRef taken = ArrayRef(In).take_while([](auto x){ return x->OrigArgIndex == Arg.getArgNo()});
In += taken.size();
return taken;
```

API wise, I think we may consider taking an input ArrayRef of elements to process, and returning two ArrayRefs -- one to the elements we've selected, and the remaining for the leftover elements to pass in on the next iteration. Might be an overkill, though. I'm just not fond of modifying external state from lambdas.

Reply:
Okay, using `take_while` as you showed above I was able to simplify further and I've removed the lambda.